src/EditorFeatures/Core/Implementation/IntelliSense/SignatureHelp/Controller_TypeChar.cs

```diff
@@ -93,56 +83,23 @@ void ICommandHandler<TypeCharCommandArgs>.ExecuteCommand(TypeCharCommandArgs arg
-                {
-                    // The user typed a character that might close the scope of the current model.
-                    // In this case, we should requery all providers.
-                    //
-                    // e.g.     Math.Max(Math.Min(1,2)$$
-                    sessionOpt.ComputeModel(allProviders, new SignatureHelpTriggerInfo(SignatureHelpTriggerReason.RetriggerCommand, triggerInfo.TriggerCharacter));
-                    computed = true;
-                }
-
-                if (textuallyTriggeredProviders.Any())
-                {
-                    // The character typed was something like "(".  It can both filter a list if
-                    // it was in a string like: Foo(bar, "(
-                    //
-                    // Or it can trigger a new list. Ask the computation to compute again.
-                    sessionOpt.ComputeModel(textuallyTriggeredProviders, untriggeredProviders, triggerInfo);
-                    computed = true;
-                }
-
-                if (!computed)
-                {
-                    // A character was typed and we haven't updated our model; do so now.
-                    sessionOpt.ComputeModel(allProviders, new SignatureHelpTriggerInfo(SignatureHelpTriggerReason.RetriggerCommand));
-                }
-            }
         }
 
-        private Tuple<List<ISignatureHelpProvider>, List<ISignatureHelpProvider>> FilterProviders(IList<ISignatureHelpProvider> providers, char ch)
+        private bool IsTextualTriggerCharacter(SignatureHelpService signatureHelpService, char ch, OptionSet options)
         {
             AssertIsForeground();
 
-            var matchedProviders = new List<ISignatureHelpProvider>();
-            var unmatchedProviders = new List<ISignatureHelpProvider>();
-            foreach (var provider in providers)
-            {
-                if (provider.IsTriggerCharacter(ch))
-                {
-                    matchedProviders.Add(provider);
-                }
-                else
-                {
-                    unmatchedProviders.Add(provider);
-                }
-            }
+            // Note: When this function is called we've already guaranteed that
+            // TypeCharWasHandledStrangely returned false.  That means we know that the caret is in
+            // our buffer, and is after the character just typed.
+
+            var caretPosition = this.TextView.GetCaretPoint(this.SubjectBuffer).Value;
+            var previousPosition = caretPosition - 1;
+            Contract.ThrowIfFalse(this.SubjectBuffer.CurrentSnapshot[previousPosition] == ch);
 
-            return Tuple.Create(matchedProviders, unmatchedProviders);
+            var trigger = SignatureHelpTrigger.CreateInsertionTrigger(ch);
+            return signatureHelpService.ShouldTriggerSignatureHelp(
+                previousPosition.Snapshot.AsText(), caretPosition, trigger, options);
```

Comment:
is there a difference between previousPosition.Snapshot and this.SubjectBuffer.CurrentSnapshot?  If not, i'd far prefer the latter.  'previousPosition' makes it sound like we're accessing something else entirely.